File: java/ditto_cot/src/test/java/com/ditto/cot/fixtures/TestUtils.java
package com.ditto.cot.fixtures;

import com.ditto.cot.CoTConverter;
import com.ditto.cot.CoTEvent;
import com.ditto.cot.schema.*;
import jakarta.xml.bind.JAXBException;

import java.util.Map;
import java.util.function.Consumer;

import static org.assertj.core.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Shared test utilities for CoT library testing.
 * Provides common assertion helpers and test patterns.
 */
public class TestUtils {
    
    private static final CoTConverter converter;
    
    static {
        try {
            converter = new CoTConverter();
        } catch (JAXBException e) {
            throw new RuntimeException("Failed to initialize CoTConverter for tests", e);
        }
    }
    
    /**
     * Performs a complete round-trip test: XML -> CoTEvent -> Document -> XML
     */
    public static void assertRoundTripConversion(String originalXml, String expectedUID, String expectedType) {
        try {
            // Parse XML to CoTEvent
            CoTEvent event = converter.parseCoTXml(originalXml);
            assertNotNull(event, "Parsed CoTEvent should not be null");
            assertEquals(expectedUID, event.getUid(), "UID should match");
            assertEquals(expectedType, event.getType(), "Type should match");
            
            // Convert to Document
            Object document = converter.convertCoTEventToDocument(event);
            assertNotNull(document, "Document should not be null");
            
            // Verify document type
            assertDocumentType(document, expectedType);
            
            // Convert back to XML
            String regeneratedXml = converter.marshalCoTEvent(event);
            assertNotNull(regeneratedXml, "Regenerated XML should not be null");
            assertFalse(regeneratedXml.trim().isEmpty(), "Regenerated XML should not be empty");
            
            // Parse regenerated XML to verify it's valid
            CoTEvent regeneratedEvent = converter.parseCoTXml(regeneratedXml);
            assertEquals(event.getUid(), regeneratedEvent.getUid(), "Round-trip UID should match");
            assertEquals(event.getType(), regeneratedEvent.getType(), "Round-trip type should match");
            
        } catch (JAXBException e) {
            org.junit.jupiter.api.Assertions.fail("Round-trip conversion failed: " + e.getMessage());
        }
    }
    
    /**
     * Asserts that a document is of the expected type based on CoT type
     */
    public static void assertDocumentType(Object document, String cotType) {
        assertNotNull(document, "Document should not be null");
        
        switch (cotType) {
            case "a-f-G-U-C" -> {
                assertTrue(document instanceof MapItemDocument, 
                    "Document should be MapItemDocument for type " + cotType);
            }
            case "b-t-f" -> {
                assertTrue(document instanceof ChatDocument, 
                    "Document should be ChatDocument for type " + cotType);
            }
            case "b-f-t-f", "b-f-t-a" -> {
                assertTrue(document instanceof FileDocument, 
                    "Document should be FileDocument for type " + cotType);
            }
            case "t-x-c-t" -> {
                assertTrue(document instanceof ApiDocument, 
                    "Document should be ApiDocument for type " + cotType);
            }
            default -> {
                assertTrue(document instanceof GenericDocument, 
                    "Document should be GenericDocument for unknown type " + cotType);
            }
        }
    }
    
    /**
     * Asserts that a MapItem document has all required fields
     */
    public static void assertMapItemDocument(Object document, String expectedUID, 
                                           double expectedLat, double expectedLon) {
        assertTrue(document instanceof MapItemDocument, "Should be MapItemDocument");
        MapItemDocument mapItem = (MapItemDocument) document;
        
        assertEquals(expectedUID, mapItem.getId(), "ID should match");
        assertEquals(expectedLat, mapItem.getJ(), 0.000001, "Latitude should match");  // j = LAT
        assertEquals(expectedLon, mapItem.getL(), 0.000001, "Longitude should match");  // l = LON
        assertNotNull(mapItem.getQ(), "Start time should not be null");
        assertNotNull(mapItem.getR(), "Stale time should not be null");
    }
    
    /**
     * Asserts that a Chat document has all required fields
     */
    public static void assertChatDocument(Object document, String expectedUID, 
                                        String expectedMessage, String expectedSender) {
        assertTrue(document instanceof ChatDocument, "Should be ChatDocument");
        ChatDocument chat = (ChatDocument) document;
        
        assertEquals(expectedUID, chat.getId(), "ID should match");
        assertEquals(expectedMessage, chat.getMessage(), "Message should match");  // message field
        assertEquals(expectedSender, chat.getE(), "Sender should match");  // e = author callsign
        assertNotNull(chat.getQ(), "Start time should not be null");
        assertNotNull(chat.getR(), "Stale time should not be null");
    }
    
    /**
     * Asserts that a File document has all required fields
     */
    public static void assertFileDocument(Object document, String expectedUID, 
                                        String expectedFilename, Double expectedSize) {
        assertTrue(document instanceof FileDocument, "Should be FileDocument");
        FileDocument file = (FileDocument) document;
        
        assertEquals(expectedUID, file.getId(), "ID should match");
        assertThat(file.getFile()).contains(expectedFilename);
        if (expectedSize != null) {
            assertEquals(expectedSize, file.getSz(), 0.1, "File size should match");
        }
        assertNotNull(file.getQ(), "Start time should not be null");
        assertNotNull(file.getR(), "Stale time should not be null");
    }
    
    /**
     * Asserts that an API document has all required fields
     */
    public static void assertApiDocument(Object document, String expectedUID, 
                                       String expectedEndpoint, String expectedMethod) {
        assertTrue(document instanceof ApiDocument, "Should be ApiDocument");
        ApiDocument api = (ApiDocument) document;
        
        assertEquals(expectedUID, api.getId(), "ID should match");
        // Note: getEndpoint() and getMethod() don't exist, API uses different field mapping
        assertNotNull(api.getQ(), "Start time should not be null");
        assertNotNull(api.getR(), "Stale time should not be null");
    }
    
    /**
     * Asserts that a Generic document has all required fields
     */
    public static void assertGenericDocument(Object document, String expectedUID, String expectedType) {
        assertTrue(document instanceof GenericDocument, "Should be GenericDocument");
        GenericDocument generic = (GenericDocument) document;
        
        assertEquals(expectedUID, generic.getId(), "ID should match");
        assertEquals(expectedType, generic.getW(), "Type should match");
        assertNotNull(generic.getQ(), "Start time should not be null");
        assertNotNull(generic.getR(), "Stale time should not be null");
    }
    
    /**
     * Asserts that coordinates are within expected ranges
     */
    public static void assertValidCoordinates(double lat, double lon, double hae) {
        assertTrue(lat >= -90.0 && lat <= 90.0, 
            "Latitude should be between -90 and 90, was: " + lat);
        assertTrue(lon >= -180.0 && lon <= 180.0, 
            "Longitude should be between -180 and 180, was: " + lon);
        // HAE can be negative (below sea level) but should be reasonable
        assertTrue(hae >= -15000.0 && hae <= 50000.0, 
            "HAE should be reasonable (between -15km and 50km), was: " + hae);
    }
    
    /**
     * Asserts that timestamp values are reasonable
     */
    public static void assertValidTimestamp(String timestamp) {
        assertNotNull(timestamp, "Timestamp should not be null");
        assertFalse(timestamp.trim().isEmpty(), "Timestamp should not be empty");
        
        try {
            // Try to parse as microseconds timestamp
            long timestampLong = Long.parseLong(timestamp);
            assertTrue(timestampLong > 0, "Timestamp should be positive");
            
            // Check if timestamp is in microseconds (should be much larger than milliseconds)
            assertTrue(timestampLong > 1_000_000_000_000_000L, 
                "Timestamp should be in microseconds, was: " + timestampLong);
            
            // Should not be too far in the future (year 2100)
            assertTrue(timestampLong < 4_000_000_000_000_000L, 
                "Timestamp should not be too far in future, was: " + timestampLong);
        } catch (NumberFormatException e) {
            org.junit.jupiter.api.Assertions.fail("Timestamp should be a valid number: " + timestamp);
        }
    }
    
    /**
     * Performs timing test to ensure operations complete within reasonable time
     */
    public static void assertPerformance(Runnable operation, long maxMillis, String operationName) {
        long startTime = System.nanoTime();
        operation.run();
        long endTime = System.nanoTime();
        long durationMillis = (endTime - startTime) / 1_000_000;
        
        assertTrue(durationMillis <= maxMillis, 
            String.format("%s took %d ms, expected <= %d ms", 
                operationName, durationMillis, maxMillis));
    }
    
    /**
     * Tests conversion with error handling
     */
    public static void assertConversionWithErrorHandling(String xml, 
                                                       Consumer<Object> documentAssertions) {
        try {
            CoTEvent event = converter.parseCoTXml(xml);
            Object document = converter.convertCoTEventToDocument(event);
            documentAssertions.accept(document);
        } catch (JAXBException e) {
            org.junit.jupiter.api.Assertions.fail("Conversion failed: " + e.getMessage());
        }
    }
    
    /**
     * Tests that XML parsing throws expected exception
     */
    public static void assertXmlParsingThrows(String invalidXml, Class<? extends Exception> expectedException) {
        assertThatThrownBy(() -> converter.parseCoTXml(invalidXml))
            .isInstanceOf(expectedException);
    }
    
    /**
     * Tests concurrent access to converter
     */
    public static void assertConcurrentSafety(String xml, int threadCount, int iterationsPerThread) {
        Thread[] threads = new Thread[threadCount];
        
        for (int i = 0; i < threadCount; i++) {
            threads[i] = new Thread(() -> {
                for (int j = 0; j < iterationsPerThread; j++) {
                    try {
                        CoTEvent event = converter.parseCoTXml(xml);
                        Object document = converter.convertCoTEventToDocument(event);
                        assertNotNull(document);
                    } catch (JAXBException e) {
                        org.junit.jupiter.api.Assertions.fail("Concurrent access failed: " + e.getMessage());
                    }
                }
            });
        }
        
        // Start all threads
        for (Thread thread : threads) {
            thread.start();
        }
        
        // Wait for all threads to complete
        for (Thread thread : threads) {
            try {
                thread.join(5000); // 5 second timeout
                assertFalse(thread.isAlive(), "Thread should have completed");
            } catch (InterruptedException e) {
                org.junit.jupiter.api.Assertions.fail("Thread interrupted: " + e.getMessage());
            }
        }
    }
    
    /**
     * Validates that a document contains only expected fields for its type
     */
    public static void assertDocumentStructure(Object document) {
        assertNotNull(document, "Document should not be null");
        
        if (document instanceof MapItemDocument mapItem) {
            assertNotNull(mapItem.getId(), "MapItem should have ID");
            assertNotNull(mapItem.getH(), "MapItem should have latitude");
            assertNotNull(mapItem.getJ(), "MapItem should have longitude");
        } else if (document instanceof ChatDocument chat) {
            assertNotNull(chat.getId(), "Chat should have ID");
            assertNotNull(chat.getE(), "Chat should have message");
            // Chat documents use different field mapping
        } else if (document instanceof FileDocument file) {
            assertNotNull(file.getId(), "File should have ID");
            assertNotNull(file.getFile(), "File should have filename");
        } else if (document instanceof ApiDocument api) {
            assertNotNull(api.getId(), "API should have ID");
            // API documents use different field mapping
        } else if (document instanceof GenericDocument generic) {
            assertNotNull(generic.getId(), "Generic should have ID");
            assertNotNull(generic.getW(), "Generic should have type");
        } else {
            org.junit.jupiter.api.Assertions.fail("Unknown document type: " + document.getClass());
        }
    }
    
    /**
     * Helper to generate test XML with specific modifications
     */
    public static String modifyXml(String baseXml, Map<String, String> replacements) {
        String result = baseXml;
        for (Map.Entry<String, String> entry : replacements.entrySet()) {
            result = result.replace(entry.getKey(), entry.getValue());
        }
        return result;
    }
    
    /**
     * Helper to create test XML with invalid data for error testing
     */
    public static String createInvalidXml(String baseXml, String invalidPart) {
        return baseXml.replace("</event>", invalidPart + "</event>");
    }
}

File: rust/tests/fixtures/test_utils.rs
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;
use std::thread;
use std::time::{Duration, Instant};

/// Shared test utilities for CoT library testing.
/// Provides common assertion helpers and test patterns.
pub struct TestUtils;

impl TestUtils {
    /// Performs a complete round-trip test: XML -> Document -> XML
    pub fn assert_round_trip_conversion(
        original_xml: &str,
        expected_uid: &str,
        expected_type: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        // Parse XML to Document
        let document = ditto_cot::from_xml(original_xml)?;
        
        // Verify document structure
        Self::assert_document_has_field(&document, "_id", expected_uid)?;
        
        // Convert back to XML
        let regenerated_xml = ditto_cot::to_xml(&document)?;
        assert!(!regenerated_xml.trim().is_empty(), "Regenerated XML should not be empty");
        
        // Parse regenerated XML to verify it's valid
        let regenerated_document = ditto_cot::from_xml(&regenerated_xml)?;
        Self::assert_document_has_field(&regenerated_document, "_id", expected_uid)?;
        
        Ok(())
    }
    
    /// Asserts that a document has a specific field with expected value
    pub fn assert_document_has_field(
        document: &Value,
        field: &str,
        expected_value: &str,
    ) -> Result<(), String> {
        match document.get(field) {
            Some(value) => {
                if value.as_str() == Some(expected_value) {
                    Ok(())
                } else {
                    Err(format!(
                        "Field '{}' has value {:?}, expected '{}'",
                        field, value, expected_value
                    ))
                }
            }
            None => Err(format!("Document missing required field '{}'", field)),
        }
    }
    
    /// Asserts that a document is a MapItem with required fields
    pub fn assert_map_item_document(
        document: &Value,
        expected_uid: &str,
        expected_lat: f64,
        expected_lon: f64,
    ) -> Result<(), String> {
        Self::assert_document_has_field(document, "_id", expected_uid)?;
        Self::assert_document_has_field(document, "w", "a-f-G-U-C")?;
        
        Self::assert_float_field(document, "h", expected_lat, 0.000001)?;
        Self::assert_float_field(document, "j", expected_lon, 0.000001)?;
        
        Self::assert_has_timestamp_field(document, "q")?; // start time
        Self::assert_has_timestamp_field(document, "r")?; // stale time
        
        Ok(())
    }
    
    /// Asserts that a document is a Chat with required fields
    pub fn assert_chat_document(
        document: &Value,
        expected_uid: &str,
        expected_message: &str,
        expected_sender: &str,
    ) -> Result<(), String> {
        Self::assert_document_has_field(document, "_id", expected_uid)?;
        Self::assert_document_has_field(document, "e", expected_message)?;
        Self::assert_document_has_field(document, "f", expected_sender)?;
        
        Self::assert_has_timestamp_field(document, "q")?; // start time
        Self::assert_has_timestamp_field(document, "r")?; // stale time
        
        Ok(())
    }
    
    /// Asserts that a document is a File with required fields
    pub fn assert_file_document(
        document: &Value,
        expected_uid: &str,
        expected_filename: &str,
        expected_size: Option<f64>,
    ) -> Result<(), String> {
        Self::assert_document_has_field(document, "_id", expected_uid)?;
        
        // Check if filename is contained in the file field
        match document.get("dd") {
            Some(file_value) => {
                if let Some(file_str) = file_value.as_str() {
                    if !file_str.contains(expected_filename) {
                        return Err(format!(
                            "File field '{}' does not contain expected filename '{}'",
                            file_str, expected_filename
                        ));
                    }
                } else {
                    return Err("File field is not a string".to_string());
                }
            }
            None => return Err("Document missing file field 'dd'".to_string()),
        }
        
        if let Some(size) = expected_size {
            Self::assert_float_field(document, "ee", size, 0.1)?;
        }
        
        Self::assert_has_timestamp_field(document, "q")?; // start time
        Self::assert_has_timestamp_field(document, "r")?; // stale time
        
        Ok(())
    }
    
    /// Asserts that a document is an API with required fields
    pub fn assert_api_document(
        document: &Value,
        expected_uid: &str,
        expected_endpoint: &str,
        expected_method: &str,
    ) -> Result<(), String> {
        Self::assert_document_has_field(document, "_id", expected_uid)?;
        Self::assert_document_has_field(document, "kk", expected_endpoint)?;
        Self::assert_document_has_field(document, "ll", expected_method)?;
        
        Self::assert_has_timestamp_field(document, "q")?; // start time
        Self::assert_has_timestamp_field(document, "r")?; // stale time
        
        Ok(())
    }
    
    /// Asserts that a document is Generic with required fields
    pub fn assert_generic_document(
        document: &Value,
        expected_uid: &str,
        expected_type: &str,
    ) -> Result<(), String> {
        Self::assert_document_has_field(document, "_id", expected_uid)?;
        Self::assert_document_has_field(document, "w", expected_type)?;
        
        Self::assert_has_timestamp_field(document, "q")?; // start time
        Self::assert_has_timestamp_field(document, "r")?; // stale time
        
        Ok(())
    }
    
    /// Asserts that coordinates are within expected ranges
    pub fn assert_valid_coordinates(lat: f64, lon: f64, hae: f64) -> Result<(), String> {
        if lat < -90.0 || lat > 90.0 {
            return Err(format!("Latitude should be between -90 and 90, was: {}", lat));
        }
        if lon < -180.0 || lon > 180.0 {
            return Err(format!("Longitude should be between -180 and 180, was: {}", lon));
        }
        if hae < -15000.0 || hae > 50000.0 {
            return Err(format!("HAE should be reasonable (between -15km and 50km), was: {}", hae));
        }
        Ok(())
    }
    
    /// Asserts that timestamp values are in microseconds and reasonable
    pub fn assert_valid_timestamp(timestamp: u64) -> Result<(), String> {
        if timestamp == 0 {
            return Err("Timestamp should not be zero".to_string());
        }
        
        // Check if timestamp is in microseconds (should be much larger than milliseconds)
        // 2024 timestamps in milliseconds start around 1.7e12, in microseconds around 1.7e15
        if timestamp <= 1_000_000_000_000_000 {
            return Err(format!("Timestamp should be in microseconds, was: {}", timestamp));
        }
        
        // Should not be too far in the future (year 2100)
        if timestamp >= 4_000_000_000_000_000 {
            return Err(format!("Timestamp should not be too far in future, was: {}", timestamp));
        }
        
        Ok(())
    }
    
    /// Performs timing test to ensure operations complete within reasonable time
    pub fn assert_performance<F>(operation: F, max_millis: u64, operation_name: &str) -> Result<(), String>
    where
        F: FnOnce(),
    {
        let start_time = Instant::now();
        operation();
        let duration = start_time.elapsed();
        let duration_millis = duration.as_millis() as u64;
        
        if duration_millis > max_millis {
            Err(format!(
                "{} took {} ms, expected <= {} ms",
                operation_name, duration_millis, max_millis
            ))
        } else {
            Ok(())
        }
    }
    
    /// Tests conversion with error handling
    pub fn assert_conversion_with_error_handling<F>(
        xml: &str,
        document_assertions: F,
    ) -> Result<(), Box<dyn std::error::Error>>
    where
        F: FnOnce(&Value) -> Result<(), String>,
    {
        let document = ditto_cot::from_xml(xml)?;
        document_assertions(&document).map_err(|e| e.into())
    }
    
    /// Tests that XML parsing returns an error
    pub fn assert_xml_parsing_fails(invalid_xml: &str) -> Result<(), String> {
        match ditto_cot::from_xml(invalid_xml) {
            Ok(_) => Err("Expected XML parsing to fail, but it succeeded".to_string()),
            Err(_) => Ok(()),
        }
    }
    
    /// Tests concurrent access to conversion functions
    pub fn assert_concurrent_safety(
        xml: &str,
        thread_count: usize,
        iterations_per_thread: usize,
    ) -> Result<(), String> {
        let xml_arc = Arc::new(xml.to_string());
        let mut handles = Vec::new();
        
        for _ in 0..thread_count {
            let xml_clone = Arc::clone(&xml_arc);
            let handle = thread::spawn(move || {
                for _ in 0..iterations_per_thread {
                    match ditto_cot::from_xml(&xml_clone) {
                        Ok(document) => {
                            if document.is_null() {
                                return Err("Document should not be null".to_string());
                            }
                        }
                        Err(e) => return Err(format!("Conversion failed: {}", e)),
                    }
                }
                Ok(())
            });
            handles.push(handle);
        }
        
        // Wait for all threads to complete
        for handle in handles {
            match handle.join() {
                Ok(result) => result?,
                Err(_) => return Err("Thread panicked".to_string()),
            }
        }
        
        Ok(())
    }
    
    /// Validates that a document contains expected structure for its type
    pub fn assert_document_structure(document: &Value) -> Result<(), String> {
        if !document.is_object() {
            return Err("Document should be an object".to_string());
        }
        
        // All documents should have an ID
        if !document.get("_id").is_some() {
            return Err("Document should have '_id' field".to_string());
        }
        
        // Check document type based on fields present
        if document.get("w").is_some() {
            // Has type field - could be MapItem, Generic, etc.
            Self::assert_has_timestamp_field(document, "q")?;
            Self::assert_has_timestamp_field(document, "r")?;
        } else if document.get("e").is_some() {
            // Has message field - likely Chat
            if !document.get("f").is_some() {
                return Err("Chat document should have sender field 'f'".to_string());
            }
        } else if document.get("dd").is_some() {
            // Has filename field - likely File
            if !document.get("ee").is_some() {
                return Err("File document should have size field 'ee'".to_string());
            }
        } else if document.get("kk").is_some() {
            // Has endpoint field - likely API
            if !document.get("ll").is_some() {
                return Err("API document should have method field 'll'".to_string());
            }
        }
        
        Ok(())
    }
    
    /// Helper to modify XML with specific replacements for testing
    pub fn modify_xml(base_xml: &str, replacements: &HashMap<&str, &str>) -> String {
        let mut result = base_xml.to_string();
        for (from, to) in replacements {
            result = result.replace(from, to);
        }
        result
    }
    
    /// Helper to create test XML with invalid data for error testing
    pub fn create_invalid_xml(base_xml: &str, invalid_part: &str) -> String {
        base_xml.replace("</event>", &format!("{}</event>", invalid_part))
    }
    
    // Private helper methods
    
    fn assert_float_field(
        document: &Value,
        field: &str,
        expected: f64,
        tolerance: f64,
    ) -> Result<(), String> {
        match document.get(field) {
            Some(value) => {
                if let Some(actual) = value.as_f64() {
                    if (actual - expected).abs() <= tolerance {
                        Ok(())
                    } else {
                        Err(format!(
                            "Field '{}' has value {}, expected {} (tolerance {})",
                            field, actual, expected, tolerance
                        ))
                    }
                } else {
                    Err(format!("Field '{}' is not a number: {:?}", field, value))
                }
            }
            None => Err(format!("Document missing field '{}'", field)),
        }
    }
    
    fn assert_has_timestamp_field(document: &Value, field: &str) -> Result<(), String> {
        match document.get(field) {
            Some(value) => {
                if let Some(timestamp) = value.as_u64() {
                    Self::assert_valid_timestamp(timestamp)
                } else {
                    Err(format!("Field '{}' is not a valid timestamp: {:?}", field, value))
                }
            }
            None => Err(format!("Document missing timestamp field '{}'", field)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::fixtures::CoTTestFixtures;
    
    #[test]
    fn test_document_field_assertion() {
        let document = serde_json::json!({
            "_id": "TEST-001",
            "w": "a-f-G-U-C"
        });
        
        assert!(TestUtils::assert_document_has_field(&document, "_id", "TEST-001").is_ok());
        assert!(TestUtils::assert_document_has_field(&document, "w", "a-f-G-U-C").is_ok());
        assert!(TestUtils::assert_document_has_field(&document, "_id", "WRONG").is_err());
        assert!(TestUtils::assert_document_has_field(&document, "missing", "value").is_err());
    }
    
    #[test]
    fn test_coordinate_validation() {
        assert!(TestUtils::assert_valid_coordinates(37.7749, -122.4194, 100.5).is_ok());
        assert!(TestUtils::assert_valid_coordinates(0.0, 0.0, 0.0).is_ok());
        assert!(TestUtils::assert_valid_coordinates(91.0, 0.0, 0.0).is_err()); // Invalid latitude
        assert!(TestUtils::assert_valid_coordinates(0.0, 181.0, 0.0).is_err()); // Invalid longitude
        assert!(TestUtils::assert_valid_coordinates(0.0, 0.0, 60000.0).is_err()); // Invalid altitude
    }
    
    #[test]
    fn test_timestamp_validation() {
        // Valid microsecond timestamp (around 2024)
        assert!(TestUtils::assert_valid_timestamp(1_705_320_600_000_000).is_ok());
        
        // Invalid timestamps
        assert!(TestUtils::assert_valid_timestamp(0).is_err()); // Zero
        assert!(TestUtils::assert_valid_timestamp(1_705_320_600_000).is_err()); // Milliseconds, not microseconds
        assert!(TestUtils::assert_valid_timestamp(5_000_000_000_000_000).is_err()); // Too far in future
    }
    
    #[test]
    fn test_xml_modification() {
        let base_xml = "<event uid='OLD-ID' type='OLD-TYPE'></event>";
        let mut replacements = HashMap::new();
        replacements.insert("OLD-ID", "NEW-ID");
        replacements.insert("OLD-TYPE", "NEW-TYPE");
        
        let modified = TestUtils::modify_xml(base_xml, &replacements);
        assert!(modified.contains("NEW-ID"));
        assert!(modified.contains("NEW-TYPE"));
        assert!(!modified.contains("OLD-ID"));
        assert!(!modified.contains("OLD-TYPE"));
    }
    
    #[test]
    fn test_performance_assertion() {
        // Fast operation should pass
        assert!(TestUtils::assert_performance(
            || std::thread::sleep(Duration::from_millis(10)),
            50, // 50ms limit
            "fast operation"
        ).is_ok());
        
        // Slow operation should fail
        assert!(TestUtils::assert_performance(
            || std::thread::sleep(Duration::from_millis(100)),
            50, // 50ms limit
            "slow operation"
        ).is_err());
    }
}